Documentation added for testing (#1281)
NVIDIA/spark-rapids: CONTRIBUTING.md

    or [help wanted](https://github.com/NVIDIA/spark-rapids/issues?q=is%3Aissue+is%3Aopen+label%3A%22help+wanted%22)
    labels.
3. Comment on the issue stating that you are going to work on it.

> 4. Code! Make sure to update unit tests and integration tests if needed! [refer to test section]

Reviewer: Suggest making the "refer to test section" a hyperlink to the section.

> 4. Code! Make sure to update unit tests and integration tests if needed! [refer to test section](#test)

Reviewer: ```suggestion
4. Code! Make sure to update unit tests and integration tests if needed! [refer to test section](#testing-your-code)
```

[Oracle Java code conventions](http://www.oracle.com/technetwork/java/codeconvtoc-136057.html)
and the Scala conventions detailed above, preferring the latter.

> ### Test

Reviewer: ```suggestion
### Testing Your Code
```

I also suggest moving this after `Sign Your Work` since that's missed by most new devs to the project and being closer to the top of the file would help.

There's enough stuff in here about testing that we may want to move it into the docs/dev directory and just link to it from this contributing document under a Testing Your Code header.

> There are two types of tests in this project. Unit tests that are written in Scala, and integration tests that are written in python. 

Reviewer: ```suggestion
There are two types of tests in this project. Unit tests that are written in Scala, and integration tests that are written in Python. 
```

We encourage writing integration tests if you have to choose between the two as that helps us ensure the plugin is working across different platforms. 

> #### Unit tests

Reviewer: The bullets in this section are not formatted properly.  See the Markdown rendering at https://github.com/NVIDIA/spark-rapids/blob/a90286e980aa74eff0191d2e496dd8ee6832724b/CONTRIBUTING.md.

In addition the bullet points are quite wordy.  If there's that much to say per bullet, I'd suggest using sub-headings with useful section header names rather than bullets.

> Unit-tests are located in com.nvidia:rapids-4-spark-tests_2.12. In order to run the unit-tests follow these steps

Reviewer: I don't think listing a maven group and artifact ID is helpful here.  Why not point them to the test Maven project in this very repo?

> 1. Build the parent project. That will build the entire project including tests

Reviewer: Similarly I think parent project is more nebulous than it needs to be here.  "Build from the top of the repository" would be clearer, IMHO.

2. Running tests

>     2a. Issue the mvn command to run the tests like 'mvn -pl com.nvidia:rapids-4-spark-tests_2.12 test'. This will run all the tests in the tests module. 

Reviewer: Why not the much more straightforward and expected `mvn test` here?  I'd rather avoid directing users to invoking specific maven projects, and it's standard to expect `mvn test` to run all the unit tests as it does today.

>     2b. To run individual tests append '-DwildcardSuites=<Fully-qualified-test-name>' to the above command e.g. To run test ParquetWriterSuite in package com.nvidia.spark.rapids, issue 'mvn test -DwildcardSuites="com.nvidia.spark.rapids.ParquetWriterSuite"'

Reviewer: Use backticks when quoting code or things the user should type verbatim.  This applies in multiple places.

> 3. Running tests against different Spark versions. There are maven profiles that should be used to run Spark version-specific tests e.g. To run tests against Spark 3.1.0, 'mvn -P spark310tests test -DwildcardSuites="com.nvidia.spark.rapids.ParquetWriterSuite"'. Please refer to the pom.xml to see the list of profiles supported

Reviewer: ```suggestion
3. Running tests against different Apache Spark versions. There are Maven profiles that should be used to run Spark version-specific tests e.g. To run tests against Spark 3.1.0, `mvn -P spark310tests test -DwildcardSuites=com.nvidia.spark.rapids.ParquetWriterSuite`. Please refer to the pom.xml to see the list of profiles supported
```

4. Spark specific configuration can be passed in by setting environment-variable SPARK_CONF 

Example: 
'SPARK_CONF="spark.rapids.memory.gpu.allocFraction=0.7,spark.sql.cache.serializer=com.nvidia.spark.rapids.shims.spark310.ParquetCachedBatchSerializer" mvn ...' this command sets the memory-allocation fraction which comes handy when dealing with OOM issues on the GPU.

#### Integration tests

> Integration tests are located in com.nvidia:rapids-4-spark-integration-tests_2.12. The suggested way to run these tests is to use the shell-script file located in the module folder called 'run_pyspark_from_build.sh'. This script takes care of some of the flags that are required to run the tests e.g. `--conf spark.driver.extraJavaOptions=-Duser.timezone=GMT --conf spark.sql.session.timeZone=UTC --conf spark.executor.extraJavaOptions=-Duser.timezone=GMT` which will have to be set for the plugin to work. It will be very useful to read the contents of the [run_pyspak_from_build.sh](https://github.com/NVIDIA/spark-rapids/blob/branch-0.3/integration_tests/run_pyspark_from_build.sh) to get better insight into what is needed as we constantly keep working on to improve and expand the plugin-support.

Reviewer: This should not refer to the Maven artifact, especially since that doesn't help the user find the scripts.  Hyperlink to the repository directory and/or the shell scripts.

Also recommend removing the details like the flags being passed, this documentation is likely to become obsolete/wrong in those areas and arguably isn't interesting/relevant at this level of documentation.

`run_pyspak_from_build.sh` is mis-spelled.

> Integration tests are located in com.nvidia:rapids-4-spark-integration-tests_2.12. The suggested way to run these tests is to use the shell-script file located in the module folder called 'run_pyspark_from_build.sh'. This script takes care of some of the flags that are required to run the tests e.g. `--conf spark.driver.extraJavaOptions=-Duser.timezone=GMT --conf spark.sql.session.timeZone=UTC --conf spark.executor.extraJavaOptions=-Duser.timezone=GMT` which will have to be set for the plugin to work. It will be very useful to read the contents of the [run_pyspak_from_build.sh](https://github.com/NVIDIA/spark-rapids/blob/branch-0.3/integration_tests/run_pyspark_from_build.sh) to get better insight into what is needed as we constantly keep working on to improve and expand the plugin-support.

Reviewer: Please word-wrap the lines being added to this file so it's friendlier to read in terminals and easier to comment on in PRs.

The tests are written python and run with pytest and the script honors pytest parameters. Some handy flags are:
    - `-k` <pytest-file-name>. This will run all the tests in that test file.
    - `-k` <test-name>. This will also run an individual test.

>     - `-s`. Doesn't capture the output and instead prints to the screen.

Reviewer: Suggest adding `-v -rfExXs` to get more details from each test run.

    - For other options and more details please visit [pytest-usage](https://docs.pytest.org/en/stable/usage.html) or type `pytest --help`

Example: 

> This command runs all the tests located in `cache_test.py` against Spark-3.1.0 using the ParquetCachedBatchSerializer and other conigs discussed above and with the debugger listening on port 5005

Reviewer: `conigs` is misspelled.

    this project or the open source license(s) involved.
```

> ### Testing your code

Reviewer: ```suggestion
### Testing Your Code
```
